flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/WaitingForResourcesTest.java

```diff
@@ -110,41 +117,187 @@ public void testNotEnoughResources() throws Exception {
+                        && !scheduledRunnable.getDelay().equals(initialResourceTimeout)
+                        && !scheduledRunnable.getDelay().equals(Duration.ZERO)) {
                     scheduledRunnable.runAction();
                 }
             }
         }
     }
 
+    private static void assertNoStateTransitionsAfterExecutingRunnables(
+            MockContext ctx, WaitingForResources wfr) {
+        Iterator<ScheduledRunnable> runnableIterator = ctx.getScheduledRunnables().iterator();
+        while (runnableIterator.hasNext()) {
+            ScheduledRunnable scheduledRunnable = runnableIterator.next();
+            if (scheduledRunnable.getExpectedState() == wfr
+                    && scheduledRunnable.getDeadline().isOverdue()) {
+                scheduledRunnable.runAction();
+                runnableIterator.remove();
+            }
+        }
+        assertThat(ctx.hasStateTransition, is(false));
+    }
+
+    private static class TestingWaitingForResources extends WaitingForResources {
+
+        private Deadline testDeadline;
+
+        TestingWaitingForResources(
+                Context context,
+                Logger log,
+                ResourceCounter desiredResources,
+                Duration initialResourceAllocationTimeout,
+                Duration resourceStabilizationTimeout) {
+            super(
+                    context,
+                    log,
+                    desiredResources,
+                    initialResourceAllocationTimeout,
+                    resourceStabilizationTimeout);
+        }
+
+        @Override
+        protected Deadline initializeOrGetResourceStabilizationDeadline() {
+            return testDeadline;
+        }
+
+        public void setTestDeadline(Deadline testDeadline) {
+            this.testDeadline = testDeadline;
+        }
+    }
+
+    @Test
+    public void testNoStateTransitionOnNoResourceTimeout() throws Exception {
+        try (MockContext ctx = new MockContext()) {
+            ctx.setHasDesiredResources(() -> false);
+            WaitingForResources wfr =
+                    new WaitingForResources(
+                            ctx,
+                            log,
+                            RESOURCE_COUNTER,
+                            Duration.ofMillis(-1),
+                            STABILIZATION_TIMEOUT);
+
+            executeAllScheduledRunnables(ctx, wfr);
+
+            assertThat(ctx.hasStateTransition(), is(false));
+        }
+    }
+
+    @Test
+    public void testStateTransitionOnResourceTimeout() throws Exception {
+        try (MockContext ctx = new MockContext()) {
+            ctx.setHasDesiredResources(() -> false);
+            WaitingForResources wfr =
+                    new WaitingForResources(
+                            ctx, log, RESOURCE_COUNTER, Duration.ZERO, STABILIZATION_TIMEOUT);
+
+            ctx.setExpectExecuting(assertNonNull());
+
+            executeAllScheduledRunnables(ctx, wfr);
+        }
+    }
+
```

Comment:
I think we are lacking a test where we have sufficient resources, then again no sufficient resources and then restart the timeout once we have sufficient resources again (e.g. testing that the sufficient resource timeout can be restarted).